### PR TITLE
test: deflake consul e2e tests

### LINKED ISSUE #7724: test flake in e2e TestConsulRegistration
A run of the nightly e2e tests failed overnight. Re-running it passed so the test appears to be possibly flaky. This test is still using gomega so it has this giant messy stack trace, but it looks like not all instances of the redis task got registered with Consul before the timeout.

```
  === RUN   TestE2E/Consul/*consul.ConsulE2ETest/TestConsulRegistration
      TestE2E/Consul/*consul.ConsulE2ETest/TestConsulRegistration: testing_t_support.go:22:
          	~/go/src/github.com/hashicorp/nomad/vendor/github.com/onsi/gomega/internal/asyncassertion/async_assertion.go:147 +0x3e9
          github.com/hashicorp/nomad/vendor/github.com/onsi/gomega/internal/asyncassertion.(*AsyncAssertion).Should(0xc00020b180, 0x168ecc0, 0xc00000f980, 0x0, 0x0, 0x0, 0x1684aa0)
          	~/go/src/github.com/hashicorp/nomad/vendor/github.com/onsi/gomega/internal/asyncassertion/async_assertion.go:48 +0x62
          github.com/hashicorp/nomad/e2e/consul.(*ConsulE2ETest).TestConsulRegistration(0xc00000e3e0, 0xc000203080)
          	~/go/src/github.com/hashicorp/nomad/e2e/consul/consul.go:75 +0x517
          reflect.Value.call(0xc00007a2a0, 0xc000206088, 0x13, 0x13a2429, 0x4, 0xc000077f20, 0x2, 0x2, 0xc000077e80, 0x40c1ca, ...)
          	/opt/goenv/versions/1.14.0/src/reflect/value.go:460 +0x8ab
          reflect.Value.Call(0xc00007a2a0, 0xc000206088, 0x13, 0xc000077f20, 0x2, 0x2, 0x1828b60, 0x1828bc0, 0xf)
          	/opt/goenv/versions/1.14.0/src/reflect/value.go:321 +0xb4
          github.com/hashicorp/nomad/e2e/framework.(*Framework).runCase.func1.2(0xc0003b47e0)
          	~/go/src/github.com/hashicorp/nomad/e2e/framework/framework.go:306 +0x25a
          testing.tRunner(0xc0003b47e0, 0xc00024e0e0)
          	/opt/goenv/versions/1.14.0/src/testing/testing.go:992 +0xdc
          created by testing.(*T).Run
          	/opt/goenv/versions/1.14.0/src/testing/testing.go:1043 +0x357

          Timed out after 5.000s.
          Expected
              <[]consul.serviceNameTagPair | len:1, cap:1>: [
                  {
                      serviceName: "redis-cache",
                      tags: {"global": {}, "cache": {}},
                  },
              ]
          to consist of
              <[]interface {} | len:1, cap:1>: [
                  [
                      {
                          serviceName: "redis-cache",
                          tags: {"global": {}, "cache": {}},
                      },
                      {
                          serviceName: "redis-cache",
                          tags: {"global": {}, "cache": {}},
                      },
                      {
                          serviceName: "redis-cache",
                          tags: {"global": {}, "cache": {}},
                      },
                  ],
              ]
```

### PR DESCRIPTION
**Pull request by schmichael**
Modernize test patterns by removing gomega and avoiding the mock_driver.

Fixes #7724

### CODE REVIEW

File: e2e/consul/consul.go [hunk 1]
```diff
@@ -29,199 +37,202 @@ func init() {
...
+	testutil.WaitForResult(func() (bool, error) {
+		consulServices, _, err := consulClient.Catalog().Service("canarytest", "", nil)
+		if err != nil {
+			return false, err
+		}
+		found := false
+		for _, s := range consulServices {
+			if e2eutil.ConsulTagsEqual(s, []string{"canary", "foo"}) {
+				found = true
+				break
```
Comment (tgross) [inline e2e/consul/consul.go:194]: Could this just be an early return? i.e. `return true, nil` and then have `return false fmt.Errorf...` at the bottom of the function

File: e2e/e2eutil/consul.go [hunk 2]
```diff
@@ -41,3 +43,11 @@ func RequireConsulDeregistered(require *require.Assertions,
 		return len(services) == 0
 	}, 5*time.Second, time.Second)
 }
+
+// ConsulTagsEqual returns true if the service has the expected tags. Both
+// slices are sorted before comparing.
+func ConsulTagsEqual(service *capi.CatalogService, expected []string) bool {
+	sort.Strings(expected)
+	sort.Strings(service.ServiceTags)
+	return reflect.DeepEqual(expected, service.ServiceTags)
```
Comment (shoenig) [inline e2e/e2eutil/consul.go:52]: seems like this could all just defer to `helper.CompareSliceSetString`

File: e2e/consul/consul.go (hunk 1 above)
Comment (shoenig) [reply e2e/consul/consul.go:194]: +1, if anything I feel `WaitForResult` breaks Go convention by not checking the error return first

File: e2e/consul/consul.go [hunk 3]
```diff
@@ -29,199 +37,202 @@ func init() {
...
-		return serviceInfo
-	}, 5*time.Second, time.Second).Should(ConsistOf([]serviceNameTagPair{
-		{"redis-cache", expectedTags},
-		{"redis-cache", expectedTags},
-		{"redis-cache", expectedTags},
-	}))
+	allocs := e2eutil.RegisterAndWaitForAllocs(f.T(), nomadClient, consulJobBasic, jobId, "")
+	require.Equal(t, 3, len(allocs))
+	allocIDs := e2eutil.AllocIDsFromAllocationListStubs(allocs)
+	e2eutil.WaitForAllocsRunning(t, tc.Nomad(), allocIDs)
```
[the remark this answers is not in the document]
Comment (schmichael) [reply e2e/consul/consul.go:68]: Yeah, that's what the failure in the issue led me to believe.

File: e2e/consul/input/canary_tags.nomad [hunk 4]
```diff
@@ -10,11 +10,25 @@ job "consul_canary_test" {
     count = 2
 
     task "consul_canary_test" {
-      driver = "mock_driver"
```
[the remark this answers is not in the document]
Comment (schmichael) [reply e2e/consul/input/canary_tags.nomad:13]: I had totally forgotten we build the mock driver into all non-release builds. Maybe we should use it more? This was the only e2e test using it, so I switched it for the sake of consistency.

**Pull request merged**